readthedocs/search/parse_json.py

```diff
@@ -1,77 +1,89 @@
 """Functions related to converting content into dict/JSON structures."""
 
-import json
 import logging
+import orjson as json
 
 from django.conf import settings
 from django.core.files.storage import get_storage_class
 
-from pyquery import PyQuery
+from selectolax.parser import HTMLParser
 
 
 log = logging.getLogger(__name__)
 
 
-def generate_sections_from_pyquery(body, fjson_storage_path):
-    """Given a pyquery object, generate section dicts for each section."""
+def generate_page_sections(body, fjson_storage_path):
+    """Generate section dicts for each section."""
 
     # Removing all <dl> tags to prevent duplicate indexing with Sphinx Domains.
+    nodes_to_be_removed = []
     try:
         # remove all <dl> tags which contains <dt> tags having 'id' attribute
-        dt_tags = body('dt[id]')
-        dt_tags.parents('dl').remove()
+        dt_tags = body.css('dt[id]')
+        for tag in dt_tags:
+            parent = tag.parent
+            if parent.tag == 'dl':
+                nodes_to_be_removed.append(parent)
     except Exception:
         log.exception('Error removing <dl> tags from file: %s', fjson_storage_path)
 
-    # remove toctree elements
+    # remove `Table of Contents` elements
     try:
-        body('.toctree-wrapper').remove()
+        nodes_to_be_removed += body.css('.toctree-wrapper') + body.css('.contents.local.topic')
     except Exception:
-        log.exception('Error removing toctree elements from file: %s', fjson_storage_path)
+        log.exception('Error removing "Table of Content" tags from file: %s', fjson_storage_path)
+
+    # removing all nodes in list
+    for node in nodes_to_be_removed:
+        node.decompose()
 
     # Capture text inside h1 before the first h2
-    h1_section = body('.section > h1')
+    h1_section = body.css('.section > h1')
     if h1_section:
-        div = h1_section.parent()
+        h1_section = h1_section[0]
+        div = h1_section.parent
         h1_title = h1_section.text().replace('¶', '').strip()
-        h1_id = div.attr('id')
+        h1_id = div.attributes.get('id', '')
         h1_content = ''
-        next_p = body('h1').next()
+        next_p = body.css_first('h1').next
         while next_p:
-            if next_p[0].tag == 'div' and 'class' in next_p[0].attrib:
-                if 'section' in next_p[0].attrib['class']:
+            if next_p.tag == 'div' and 'class' in next_p.attributes:
+                if 'section' in next_p.attributes['class']:
                     break
 
-            text = parse_content(next_p.text(), remove_first_line=True)
+            text = parse_content(next_p.text(), remove_first_line=False)
```

Comment:
Also here. It seems we are changing the logic.

Reply:
Yes. During working on it again, I find that few lines were getting removed from indexing because of this.

Reply:
Should I target these things in another PR?

Reply:
I would say "yes".

If we first apply the logic changes and after that we apply the migration, we can easily come back to the previous (pyquery, etc) implementation with just one revert.